superduperdb/db/base/cdc.py

```diff
@@ -45,39 +327,113 @@ class DatabaseListenerFactory(t.Generic[DBListenerType]):
+    def __init__(self, db: 'DB'):
+        self.db = db
+        self._cdc_stop_event = Event()
+        self.CDC_QUEUE: queue.Queue = queue.Queue()
+        self.cdc_change_handler: t.Optional[CDCHandler] = None
+        self._CDC_LISTENERS: t.Dict[str, BaseDatabaseListener] = {}
+        self._running: bool = False
+        self._cdc_existing_collections: t.MutableSequence['TableOrCollection'] = []
+
+    @property
+    def running(self) -> bool:
+        return self._running
+
+    def start(self):
+        """
+        This method starts the cdc process on the database.
+        """
+        self._running = True
+
+        # listen to existing collection without cdc enabled
+        list(map(lambda on: self.listen(on), self._cdc_existing_collections))
+
+    def listen(
+        self,
+        on: 'TableOrCollection',
+        identifier: str = '',
+        *args,
+        **kwargs,
+    ):
+        """
+        Starts cdc service on the provided collection
+        Not to be confused with ``superduperdb.container.listener.Listener``.
+
+        :param db: A superduperdb instance.
+        :param on: Which collection/table listener service this be invoked on?
+        :param identifier: A identity given to the listener service.
+        """
+        from superduperdb.db.base import backends
+
+        if isinstance(self.db.databackend, backends.MongoDataBackend):
+            db_type = 'mongodb'
+        else:
+            raise NotImplementedError(f'{self.db.databackend} not supported yet!')
+
+        if not self.cdc_change_handler:
+            cdc_change_handler = CDCHandler(
+                db=self.db, stop_event=self._cdc_stop_event, queue=self.CDC_QUEUE
+            )
+            cdc_change_handler.start()
+            self.cdc_change_handler = cdc_change_handler
+
+        db_factory: DatabaseListenerFactory = DatabaseListenerFactory(db_type=db_type)
+        listener = db_factory.create(
+            db=self.db, on=on, identifier=identifier, *args, **kwargs
+        )
+        self._CDC_LISTENERS[on.identifier] = listener
+
+        listener.listen()
+        return listener
+
+    def stop(self):
+        """
+        Stop all registered listeners
+        """
+        for _, listener in self._CDC_LISTENERS.items():
+            listener.stop()
+
+    def stop_handler(self):
+        """
+        Stop the cdc handler thread
+        """
+        self._cdc_stop_event.clear()
+        if self.cdc_change_handler:
+            self.cdc_change_handler.join()
 
-    factory_factory = DatabaseListenerFactory[MongoDatabaseListener]
-    db_factory = factory_factory(db_type=db_type)
-    return db_factory.create(db=db, on=on, identifier=identifier, *args, **kwargs)
+    def add(self, collection: 'TableOrCollection'):
+        """
+        This method registered the given collection for cdc
+        """
+        if self.running:
+            self.listen(collection)
```

Comment:
I don't see append to existing `_cdc_exists_collections` if `self.running` is `True`.